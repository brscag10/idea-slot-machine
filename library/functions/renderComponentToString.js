import Vue from "vue";
import { createRenderer } from "vue-server-renderer";

/**
 * create an html string from a Vue Component for testing purposes
 *
 * @export
 * @param {VueComponent} VueComponent the vue component we wish to render
 * @returns {Promise<string>} the rendered component string
 */
export default async function renderComponentToString(VueComponent, props = {}) {
  const renderer = createRenderer();

  const ComponentWithProps = Vue.extend(VueComponent);
  const vm = new ComponentWithProps(props).$mount();

  return renderer.renderToString(vm);
}
