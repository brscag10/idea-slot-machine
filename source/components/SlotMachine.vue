<template>
<section class="SlotMachine" :style="{ grid: `auto / repeat(${slots.length}, 1fr)` }">
  <slot-wheel v-for="{ name, options } in slots"
    ref="wheels"
    :key="name"
    :name="name"
    :options="options"
  />
</section>
</template>

<script lang="ts">
import Component from "vue-class-component";
import SlotWheel from "./SlotWheel.vue";
import Vue from "vue";

@Component({
  components: {
    SlotWheel
  },
  props: {
    slots: {
      default: () => [],
      type: Array
    }
  }
})
/**
 * SlotMachine holds all the Wheels
 */
export default class SlotMachine extends Vue {
  /**
   * `spin` - spins all the wheels
   * 
   * @returns {void}
   */
  async spin() {
    await Promise.all(
      (this.$refs.wheels as SlotWheel[]).map(
        wheel => wheel.spin()
      )
    );

    this.$emit("done");
  }
};
</script>

<style scoped>
.SlotMachine {
  display: grid;
  grid-column-gap: var(--horiz-padding);

  width: 100%;
  max-width: var(--max-supported-width);
  min-width: var(--min-supported-width);

  padding: var(--vert-padding) var(--horiz-padding);
}
</style>
