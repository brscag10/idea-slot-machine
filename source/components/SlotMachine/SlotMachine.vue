<template>

</template>

<style scoped>
.SlotMachine {
  display: grid;
  grid-column-gap: var(--horiz-padding);

  width: 100%;
  max-width: var(--max-supported-width);
  min-width: var(--min-supported-width);

  padding: var(--vert-padding) var(--horiz-padding);
}
</style>
