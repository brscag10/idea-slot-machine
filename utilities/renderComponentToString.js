import Vue from "vue";
import { createRenderer } from "vue-server-renderer";

/**
 * create an html string from a Vue Component for testing purposes
 *
 * @export
 * @param {VueComponent} VueComponent the vue component we wish to render
 * @returns {Promise<string>} the rendered component string
 */
export default async function renderComponentToString(VueComponent) {
  const renderer = createRenderer();
  const v = new Vue({
    el: document.createElement("div"),
    render: h => h(VueComponent)
  });

  return new Promise((resolve, reject) => {
    renderer.renderToString(v, (error, result) => {
      if (error) {
        reject(error);
      }

      resolve(result);
    });
  });
}
