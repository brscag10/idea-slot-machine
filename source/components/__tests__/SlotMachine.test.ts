import SlotMachine from "../SlotMachine.vue";
import renderComponentToString from "../../../utilities/renderComponentToString";

test.concurrent("passes a smoke test", async (): Promise<void> => {
  const result = await renderComponentToString(SlotMachine);

  expect(result).toMatchSnapshot();
});
