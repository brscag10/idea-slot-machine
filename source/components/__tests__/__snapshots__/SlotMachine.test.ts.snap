// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`passes a smoke test 1`] = `"<section data-server-rendered=\\"true\\" class=\\"SlotMachine\\" style=\\"grid:auto / repeat(0, 1fr);\\"></section>"`;
